### ISSUE TITLE
`ttnn.concat` crashes in `apply_descriptor_runtime_args` after descriptor migration when invoked back-to-back with different shapes

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-metal#45089]
Regression found in Forge models, plz fix, here are some claude generated details.

## Summary

After descriptor-style migration of `ttnn.concat` (PR #43535, in `origin/main`), running an interleaved-DRAM tiled BF16 `ttnn.concat` on the last dim and then a **larger** interleaved-DRAM tiled BF16 `ttnn.concat` on the last dim within the same process — with program cache enabled (default) — crashes the second invocation with:

```
TT_FATAL @ tt_metal/impl/kernels/kernel.cpp:572:
    logical_core.x < this->core_to_runtime_args_.size() &&
    logical_core.y < this->core_to_runtime_args_[logical_core.x].size()
Cannot get runtime args for kernel reader_concat_interleaved_start_id
that is not placed on core 0-4
```

Disabling the program cache (`device.disable_and_clear_program_cache()`) makes the failure go away.

## Minimal repro (14 lines)

```python
import torch
import ttnn

def to_tt(t, device):
    return ttnn.from_torch(
        t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )

def main():
    device = ttnn.open_device(device_id=0)
    try:
        # 1) decode-shape concat — caches a program
        a = torch.rand(1, 1, 1, 64, dtype=torch.bfloat16)
        tt_a = to_tt(a, device)
        ttnn.concat([tt_a, tt_a], dim=3, memory_config=ttnn.DRAM_MEMORY_CONFIG)

        # 2) prefill-shape concat — CRASH on this one
        b = torch.rand(1, 4, 64, 64, dtype=torch.bfloat16)
        c = torch.rand(1, 4, 64, 64, dtype=torch.bfloat16)
        tt_b, tt_c = to_tt(b, device), to_tt(c, device)
        ttnn.concat([tt_b, tt_c], dim=3, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    finally:
        ttnn.close_device(device)

main()
```

Run on a single Wormhole device (n300 / n150). Reproduces 100%.

Output (current `origin/main`):
```
small concat ok, output shape=(1, 1, 1, 128)
cache entries: 1
large concat:
  TT_FATAL: Cannot get runtime args for kernel
  reader_concat_interleaved_start_id that is not placed on core 0-4
```

Disabling the cache before the second concat prevents the crash:
```python
device.disable_and_clear_program_cache()  # add this between (1) and (2)
```

## Stack trace

```
tt::tt_metal::Kernel::runtime_args_data(tt::xy_pair const&)
tt::tt_metal::GetRuntimeArgs(Program const&, unsigned, tt::xy_pair const&)
tt::tt_metal::apply_descriptor_runtime_args(Program&, ProgramDescriptor const&)
  ← ttnn/api/ttnn/mesh_device_operation_adapter.hpp:553
ttnn::device_operation::detail::handle_mesh_adapter_cache_hit<
    DeviceOperationWithMeshDeviceAdapter,
    MeshDeviceOperationAdapter<ConcatDeviceOperation>>
ttnn::device_operation::detail::launch_operation_with_adapter<…>
ttnn::prim::concat(…)
ttnn::operations::data_movement::concat_impl(…)
```

## Diagnosis (working theory)

In `ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp`, `ConcatProgramFactory::create_descriptor` chooses the number of cores via `split_work_to_cores(compute_with_storage_grid_size, num_output_pages, …)`. `num_output_pages` is a function of output shape, so different output shapes produce **different `all_cores` sets** (and the reader/writer kernels are placed only on those cores).

In `mesh_device_operation_adapter.hpp::apply_descriptor` (the cache-hit path, lines 502–558), because the migrated factory uses the *old-style* `runtime_args.emplace_back(core, std::move(args))` and does **not** declare `emplace_runtime_args()` buffer bindings, the fast-path `apply_resolved_bindings` branch is skipped and the code falls into:

```cpp
auto desc = invoke_per_coord(attrs, tensor_args, tensor_return_value, ...);
tt::tt_metal::apply_descriptor_runtime_args(program, desc);
```

`invoke_per_coord` re-runs `create_descriptor` on the **new** tensor args, producing a new `desc` whose `kernel.runtime_args` is indexed by the **new** shape's cores. `apply_descriptor_runtime_args` then iterates those cores and calls `GetRuntimeArgs(cached_program, k, core)` — but `cached_program`'s kernel was only placed on the **old** shape's cores. When the new shape's core set includes a core that the old set did not (e.g. `(0, 4)`), `runtime_args_data` asserts.

Equivalent informal statement: the descriptor cache-hit path assumes the factory is shape-invariant in its kernel placement. For ops whose `split_work_to_cores` depends on output size, that assumption is broken once shape varies between invocations.

## Pre-migration behavior (why this didn't surface before)

Before PR #43535, the cache-hit path used `override_runtime_arguments`, which iterated `shared_vars.cores` (the FIRST invocation's cores) and only overwrote address fields. With different shapes, this silently produced incorrect outputs (wrong work distribution) but did not crash. The descriptor migration replaced that with a re-derived descriptor + apply, which is correct *only* when the freshly-derived core set equals the cached kernel placement — surfacing the latent shape-dependence as a hard assertion.

## Suggested fix directions (in order of preference)

1. **Force cache-key sensitivity to shape** — add a `compute_program_hash` override on `ConcatDeviceOperation` that mixes in `input_tensor.logical_shape()` and `output_tensor.padded_shape()` / `num_output_pages` so two different-shape invocations always cache-miss. Cheap, correct, no factory rewrite.
2. **Make `ConcatProgramFactory` shape-invariant in core placement** — always use the full compute grid (no `split_work_to_cores` fan-in), keying per-core work off runtime args. Bigger change; questionable perf.
3. **Migrate `ConcatProgramFactory` to declare `emplace_runtime_args()` buffer bindings** — that takes the fast `apply_resolved_bindings` branch on cache hits, which doesn't re-derive the descriptor. This still requires that the factory's core placement be stable across invocations, so (1) is a precondition unless you also do (2).
4. **Detect mismatch and invalidate** — in `apply_descriptor`, compare the freshly-derived core set against the cached program's kernel placement; on mismatch, evict the cache entry and treat as miss. Defensive but masks the underlying hashing/contract bug.

## Real-world impact (downstream context for prioritization)

Found on a tt-xla PR qualifying a tt-mlir uplift against a previously-working vLLM model: `Qwen/Qwen3-0.6B` under tensor parallel on `n300_llmbox`. The uplift moves RoPE fusion to TTIR (tt-mlir `06fd815f1`), whose TTNN decomposition fallback ends up emitting the small `1x1x1x64 → 1x1x1x128` and larger `1x4x64x64 → 1x4x64x128` last-dim `ttnn.concat` ops in the same compiled program — the pair that trips this issue. The bug isn't RoPE-specific; any flow that calls `ttnn.concat` on the last dim with progressively larger tiled interleaved-DRAM tensors in one process will hit it.

## Run environment

- tt-metal: commit `c5ebc635109` (2026-05-19), which includes the descriptor migration `438eaf0d816` (2026-05-11). Bug is also present in current `origin/main`.
- Device: Wormhole (single chip, n300 reservation)
- Layout: tile, BF16, DRAM interleaved on both inputs and output
- Dim: last (3 for 4-D), confirmed; have not checked other dims

## Workaround tested

Setting `TT_RUNTIME_ENABLE_PROGRAM_CACHE=0` (tt-xla env var) eliminates the crash. Disabling the cache directly via `device.disable_and_clear_program_cache()` in standalone TTNN tests has the same effect.

## Extras

**Suggested assignees / mention:** @martemovTT (author of #43535 descriptor migration) and @tenstorrent/metalium-developers-ops-data-movement  (CODEOWNERS for `ttnn/cpp/ttnn/operations/data_movement/`)

**Related:**
- #43535 — `migrate ttnn::concat to the descriptor style` (likely-introducing PR)
- #42213 — `[Descriptor Migration] data_movement/concat` (tracking issue, closed when #43535 landed)
- #44294 — `Remove dead override_runtime_arguments() from matmul factories using create_descriptor()` (same migration pattern, different op)
- #43838 — `Optimize descriptor framework cache-hit fast path: reduce per-binding GetRuntimeArgs overhead` (related framework path)

**Comment by kmabeeTT** [comment]
> Sharing additional findings from chasing this downstream — the failure mode is broader than the original repro suggests.

## Root cause is the hash function, not the concat factory

`ConcatDeviceOperation` falls back to the default `hash_objects_with_default_seed` (no custom `compute_program_hash`). That walks the operation attributes and `tensor_args.input_tensors`, ultimately hashing each input shape as a sequence of `uint32` dim values through `tt_stl::hash::hash_objects`, which is a port of `boost::hash_combine`:

```cpp
// tt_stl/tt_stl/reflection.hpp
inline hash_t hash_objects(hash_t seed, const Types&... args) noexcept {
    ([&seed](const auto& arg) {
        seed ^= hash_object(arg) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
    }(args), ...);
    return seed;
}
```

For integer dims `hash_object(int x) = x`, so the shape hash reduces to:
```
seed = 0
for d in dims: seed ^= d + 0x9e3779b9 + (seed << 6) + (seed >> 2)   // 64-bit
```

This combiner is known to have weak distribution on small integer inputs (this is why Boost's combiner has been a recurring source of cache-key collisions in other projects too).

## The collision is systematic, not a one-off

The original repro pair `(1, 1, 1, 64)` ↔ `(1, 4, 64, 64)` colliding looked accidental, but a Python re-implementation of the exact combiner reveals that the **3-element prefixes** `[1, 1, 1]` and `[1, 4, 64]` hash to the same intermediate seed. Once two prefixes collide, **any common suffix yields a colliding full-shape hash**:

| `D` | hash `(1, 1, 1, D)` | hash `(1, 4, 64, D)` | collide |
|---|---|---|---|
| 16 | `0x00028253c81abb68` | `0x00028253c81abb68` | yes |
| 32 | `0x00028253c81abb78` | `0x00028253c81abb78` | yes |
| 64 | `0x00028253c81abb58` | `0x00028253c81abb58` | yes |
| 128 | `0x00028253c81abb18` | `0x00028253c81abb18` | yes |
| 256 | `0x00028253c81ab898` | `0x00028253c81ab898` | yes |

(verified bit-for-bit against the C++ combiner)

A short sweep over realistic RoPE-relevant shape combinations (head_dim ∈ {16…512}, kv_heads ∈ {1…16, 32, 64}, seq ∈ {1…4096}) finds **90+ distinct-shape collisions**. A representative sample of pairs that all hash equal:

```
(1, 1, 1, D)   ↔  (1, 4, 64, D)        for D ∈ {16, 32, 40, 48, 64, 80, 96, 128, 160, 192, 256}
(1, 5, 64, D)  ↔  (1, 6, 1, D)         for many D
(1, 7, 64, D)  ↔  (1, 8, 1, D)
(1, 10, 128, D) ↔ (1, 16, 1, D)
(1, 12, 384, D) ↔ (1, 14, 1, D)
(1, 14, 2, D)  ↔  (1, 16, 128, D)
(1, 4, 384, D) ↔  (1, 8, 128, D+64)    (head_dim cross-product)
(1, 4, 512, D) ↔  (1, 8, 256, D+64)
(1, 4, 768, D) ↔  (1, 8, 512, D+64)
(1, 4, 1024, D) ↔ (1, 8, 768, D+64)
```

These aren't just theoretical — `(1, 1, 1, 64)` ↔ `(1, 4, 64, 64)` is the live pair that crashes the Qwen3-0.6B vLLM run; the others are all shapes any GQA decoder LLM could plausibly produce in a single compiled program under tensor-parallel sharding.

## Implications for the fix

1. **A custom `compute_program_hash` on `ConcatDeviceOperation` that includes input shapes through `hash_objects` won't help** — the underlying combiner is the weak link.
2. **Adding a discriminator that breaks shape-prefix equivalence is what's needed.** Options off the top of my head:
   - hash `shape.size()` first (rank) before the dims, so `[1, 1, 1, D]` and `[1, 4, 64, D]` start with different seeds (this alone wouldn't fix it — both ranks are 4 — but mixing in `volume()` or `padded_shape()` would).
   - hash `padded_shape()` or per-dim tile counts instead of raw logical dims.
   - swap `hash_objects` for a stronger combiner (xxhash / wyhash) at least for shape data.
3. **Alternatively**, fix the cache-hit handler to *validate* that the cached program's kernel core_ranges cover the newly-derived descriptor's `runtime_args` coreset, and treat a mismatch as cache eviction rather than `TT_FATAL`. That's defensive, doesn't fix root cause, but bounds the blast radius — turns "silent wrong-result risk under override path" or "hard-crash under descriptor path" into "extra cache miss."

The tt-mlir-side fix in [tenstorrent/tt-mlir#8579](https://github.com/tenstorrent/tt-mlir/issues/8579) only removes the *specific* `(1, 1, 1, head_dim/2)` shape from being emitted by RoPE decomposition — it doesn't address the broader weakness.

## How to reproduce the systematic collision

`tt_stl::hash::hash_objects` re-implemented in Python:

```python
MASK64 = (1 << 64) - 1
def shape_hash(dims):
    seed = 0
    for d in dims:
        seed = (seed ^ ((d + 0x9e3779b9 + ((seed << 6) & MASK64) + (seed >> 2)) & MASK64)) & MASK64
    return seed

assert shape_hash([1, 1, 1, 64]) == shape_hash([1, 4, 64, 64])
# 0x00028253c81abb58 in both cases
```

Round-trip verification against the TTNN runtime (program-cache hit observed via `device.num_program_cache_entries()` staying constant across the supposedly-distinct calls) confirms this is the same key the program cache uses.
